src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs

```diff
@@ -1297,6 +1306,94 @@ private BoundExpression RewriteDecimalConversion(BoundConversion oldNode, CSharp
             }
         }
 
+        private BoundExpression RewriteMethodGroupConversion(BoundConversion oldNode, CSharpSyntaxNode syntax, BoundExpression operand, MethodSymbol targetMethod, bool isExtensionMethod, TypeSymbol type)
+        {
+            Debug.Assert(type.IsDelegateType());
+            Debug.Assert((object)targetMethod != null);
+
+            var cacheContainer = ObtainCacheContainer(type, targetMethod);
+            var cacheField = cacheContainer.ObtainCacheField(_factory, type, targetMethod);
+
+            if (cacheContainer.ContainerKind == DelegateCacheContainerKind.MethodScopedGeneric)
+            {
+                cacheField = cacheField.AsMember(cacheContainer.Construct(_factory.TopLevelMethod.TypeParameters));
+            }
+
+            var orgSyntax = _factory.Syntax;
+            _factory.Syntax = syntax;
+
+            var boundCacheField = _factory.Field(null, cacheField);
+            var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, operand, targetMethod, isExtensionMethod, type)
+            {
+                WasCompilerGenerated = true
+            };
+
+            var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+            _factory.Syntax = orgSyntax;
+
+            return rewrittenNode;
+        }
+
+        private DelegateCacheContainer ObtainCacheContainer(TypeSymbol delegateType, MethodSymbol targetMethod)
+        {
+            var currentMethod = _factory.TopLevelMethod;
+            var containerKind = ChooseDelegateCacheContainerKind(currentMethod, delegateType, targetMethod);
+
+            if (containerKind == DelegateCacheContainerKind.ModuleScopedConcrete)
+            {
+                var compilation = _factory.Compilation;
+                return compilation.DelegateCacheManager.ObtainModuleScopedContainer(compilation, delegateType);
+            }
+
+            if (containerKind == DelegateCacheContainerKind.TypeScopedConcrete)
+            {
+                var typeCompilationState = _factory.CompilationState;
+                var tsContainer = typeCompilationState.TypeScopedDelegateCacheContainerOpt;
+                if ((object)tsContainer == null)
+                {
+                    typeCompilationState.TypeScopedDelegateCacheContainerOpt
+                        = tsContainer
+                        = new TypeOrMethodScopedDelegateCacheContainer(currentMethod.ContainingType, _factory.ModuleBuilderOpt.CurrentGenerationOrdinal);
+                    _factory.AddNestedType(tsContainer);
+                }
+                return tsContainer;
+            }
+
+            var msContainer = _methodScopedDelegateCacheContainer;
```

Comment:
Add assert ```containerKind == DelegateCacheContainerKind.MethodScopedGeneric```. Or better use a switch statement over containerKind, with default case throwing unreachable exception, instead of a sequence of ```if``` statements.